proposals/XXXX-recursive-protocol-constraints.md

```diff
@@ -0,0 +1,265 @@
+relatively simple change. It is only a new feature in the sense that certain associated type definitions which were
+previously disallowed will now be accepted by the compiler.
+
+Implementation details regarding the compiler changes necessary to implement the first part of the solution can be
+found in [this document](https://gist.github.com/DougGregor/e7c4e7bb4465d6f5fa2b59be72dbdba6).
+
+The second part of the solution involves updating the standard library to take advantage of the removal of this
+restriction. Such changes are made with [SE-0142](https://github.com/apple/swift-evolution/blob/master/proposals/0142-associated-types-constraints.md)
+in mind, and incorporate both recursive constraints and `where` clauses. The changes necessary for this are described
+in the _Detailed Design_ section below.
+
+This second change will affect the sort of user code which is accepted by the compiler. User code which uses the
+affected protocols and types will require fewer generic parameter constraints to considered valid. Conversely,
+user code which (incorrectly) uses the private protocols removed by this proposal, or which uses the affected public
+protocols in an incorrect manner, might cease to be accepted by the compiler after this change is implemented. 
+
+## Detailed design
+
+The following standard library types and protocols will be removed:
+
+* `_BidirectionalIndexable`
+* `_Indexable`
+* `_IndexableBase`
+* `_MutableIndexable`
+* `_RandomAccessIndexable`
+* `_RangeReplaceableIndexable`
+
+The following standard library protocols and types will change.
+
+Note that since the specific collection types conform to `Collection`, and `Collection` refines `Sequence`, not all the
+constraints need to be defined on every collection-related associated type.
+
+Default values of all changed associated types remain the same, unless explicitly noted otherwise.
+
+All "Change associated type" entries reflect the complete, final state of the associated type definition, including
+removal of underscored protocols and addition of any new constraints.
+
+### `Any*Collection` (all variants)
+
+* Remove all `C.Subsequence` and `C.Indices` constraints from public `init`
+
+### `Arithmetic`
+
+* Change associated type: `associatedtype Magnitude : Arithmetic`
+
+### `BidirectionalCollection`
+
+* Remove conformance to `_BidirectionalIndexable`
+* Change associated type: `associatedtype SubSequence : BidirectionalCollection`
+* Change associated type: `associatedtype Indices : BidirectionalCollection`
+
+### `Collection`
+
+* Remove conformance to `_Indexable`
+* Change associated type: `associatedtype SubSequence : Collection where SubSequence.Index == Index, SubSequence.Indices == Indices`
+* Change associated type: `associatedtype Indices : Collection where Indices.Iterator.Element == Index, Indices.Index == Index, Indices.SubSequence == Indices`
```

Comment:
I think I added this constraint because I thought it was necessary for some implementation detail of the generic `Slice` type.